### PR TITLE
Adds Taxonomy, Tag, ObjectTag models and APIs

### PR DESCRIPTION
**Pull request by pomegranited**
## Description

Implements the Taxonomy, Tag, and ObjectTag models, APIs, and permissions described in the [discovery document](https://docs.google.com/document/d/13zfsGDfomSTCp_G-_ncevQHAb4Y4UW0d_6N8R2PdHlA/edit?usp=sharing) for the tagging app.

## Supporting information

Relates to https://github.com/openedx/modular-learning/issues/63

## Testing instructions

Manual testing is possible but not very useful, since we don't yet have views for these models and APIs.

* Ensure that all models and methods from the discovery doc are implemented.
* Ensure that the tests cover the expected behavior of the models, APIs and permissions as described in the discovery doc.

## Other information

The `import_tags`, `export_tags` and `autocomplete_tags` will be added by a subsequents PRs

SystemTaxonomy will be added by a subsequent PR as a subclass of Taxonomy.

ContentTaxonomy and ContentTag will subclass these models, and will be implemented in edx-platform `openedx.features.content_tagging`.

## To do before merge

- [x] use multi-collation field types added by https://github.com/openedx/openedx-learning/pull/51
- [x] squash migrations into a single `0001_initial.py`
- [ ] squash and merge this change (don't just merge)

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @pomegranited!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

File: openedx_tagging/core/tagging/models.py
Comment (bradenmacdonald) [inline openedx_tagging/core/tagging/models.py:57]: I don't think this unique constraint will work as expected when `parent_id` is NULL, because null values are always different from each other. So you can have multiple entries with the same value and a NULL parent in the same taxonomy.

I think you can achieve what you want using [`UniqueConstraint`](https://docs.djangoproject.com/en/4.2/ref/models/constraints/#uniqueconstraint) which is more flexible.

**Review by bradenmacdonald — COMMENTED** [review]
Very cool :) I am liking this approach.

File: openedx_tagging/core/tagging/models.py
Comment (bradenmacdonald) [inline openedx_tagging/core/tagging/models.py:138]: Can you remind me, what is the use case for being able to override `ObjectTag` per taxonomy?

**Comment by ChrisChV** [comment]
>Other information
>
>SystemTaxonomy will be added by a subsquent PR as a subclass of Taxonomy.
>
>ContentTaxonomy and ContentTag will subclass these models, and will be implemented in edx-platform openedx.features.tagging.

Also the `import_tags`, `export_tags` and `autocomplete_tags` will be added by a subsequents PRs

**Comment by ChrisChV** [comment]
@pomegranited Great work! Looks good :+1:

File: openedx_tagging/core/tagging/models.py
Comment (bradenmacdonald) [reply openedx_tagging/core/tagging/models.py:138]: I generally like the YAGNI philosophy so I'd say add it when needed :)

Comment (pomegranited) [reply openedx_tagging/core/tagging/models.py:138]: Cool.. I'll put this code into the discovery in case it's needed later, but will remove it from here.

File: openedx_tagging/core/tagging/models.py
Comment (pomegranited) [reply openedx_tagging/core/tagging/models.py]: Have removed this method.

File: openedx_tagging/core/tagging/models.py
Comment (pomegranited) [reply openedx_tagging/core/tagging/models.py:477]: Removed the object_tag_class property, but noted this ^ in [the discovery](https://docs.google.com/document/d/13zfsGDfomSTCp_G-_ncevQHAb4Y4UW0d_6N8R2PdHlA/edit#heading=h.brvhx3mc8b5) if we end up needing it later.

File: docs/decisions/0009-tagging-administrators.rst
Comment (ChrisChV) [reply docs/decisions/0009-tagging-administrators.rst:29]: @pomegranited  It's reasonable to me. Maybe it is good to add this also as a comment in the API so that it is taken into account when using it

File: openedx_tagging/core/tagging/api.py
Comment (bradenmacdonald) [inline openedx_tagging/core/tagging/api.py:66]: If there are a _lot_ of `ObjecTag` rows, this has the downside of loading them all from the database and then converting them all into python objects before iterating over them. It seems like using `for object_tag in ObjectTag.objects.all().iterator()` [is better](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#iterator):

> For a `QuerySet` which returns a large number of objects that you only need to access once, this can result in better performance and a significant reduction in memory.

However, it's still not as good as I hoped:

> MySQL doesn’t support streaming results, hence the Python database driver loads the entire result set into memory. 

What?! Yet another point for postgres...

File: openedx_tagging/core/tagging/models.py
Comment (pomegranited) [inline openedx_tagging/core/tagging/models.py:24]: @ormsbee I'm not clear on when it's better to use an `id` vs a `uuid`, and so I'm not sure which is best as the primary key for these models: Tag, Taxonomy, ObjectTag.

Since we're recommending multi-table inheritance to extend these models, they will share a database. However if we ever move this app to a separate service, I can see how a separate database would prefer to store UUIDs.

What do you think?

File: openedx_tagging/core/tagging/api.py
Comment (pomegranited) [reply openedx_tagging/core/tagging/api.py:66]: @bradenmacdonald Cool, I didn't know about that! Will fix this.

>> MySQL doesn’t support streaming results, hence the Python database driver loads the entire result set into memory.
> 
> What?! Yet another point for postgres...

My favourite related comment so far came from this issue: https://code.djangoproject.com/ticket/2495#comment:14

>  After eight years, and considering that the fix is far from easy, could we file this under "use a better database" and close the ticket?

**Comment by pomegranited** [comment]
@ChrisChV My apologies for my [rebase on `main`](https://github.com/openedx/openedx-learning/compare/71a84ce64db0f126c2d763cb43a11a243857041d..c37f9f4ffdf74e9967d37f5057f307fd28595278).. I've rolled it back and did a [merge with `main`](https://github.com/openedx/openedx-learning/pull/57/commits/fbbd6f60744f4e62c9be61dddf9b16f87e1b41bb) instead, so you should still be able to change the base branch of your open PRs to this one to ease reviews there.

But I'll need to do a rebase before I merge here so I can squash those migrations. Awaiting word on [whether we need a uuid field](https://github.com/openedx/openedx-learning/pull/57#discussion_r1236025461) before I do that.

File: openedx_tagging/core/tagging/models.py
Comment (ormsbee) [reply openedx_tagging/core/tagging/models.py:26]: I think it's okay to have just an `id` here, since the combination of Taxonomy and `external_id` should be enough to uniquely identify a tag. If we find a need for UUIDs, we can add them later.

File: openedx_tagging/core/tagging/models.py
Comment (ormsbee) [inline openedx_tagging/core/tagging/models.py:60]: This style is deprecated in 4.2. Please use [indexes](https://docs.djangoproject.com/en/4.2/ref/models/options/#django.db.models.Options.indexes) instead.

File: openedx_tagging/core/tagging/models.py
Comment (ormsbee) [inline openedx_tagging/core/tagging/models.py:50]: Should there be a unique constraint on `(taxonomy, external_id)`?

File: openedx_tagging/core/tagging/models.py
Comment (ormsbee) [inline openedx_tagging/core/tagging/models.py:7]: Should we move the fields stuff out of `openedx_learning` and into something else? I'm just worried about circular dependencies here, since my understanding is that eventually there will something in `openedx_learning.contrib` that imports from `openedx_tagging` to link tags to the content data models?

File: openedx_tagging/core/tagging/api.py
Comment (ormsbee) [inline openedx_tagging/core/tagging/api.py:68]: I realize I've been a bit in-and-out of the taxonomy/tagging model discussions, so please pardon me if you folks have already gone over this:

I wonder if this could be less expensive if we make one more layer of model indirection between the Object and the tag. Right now, it seems like we have the global representation of a Tag, and a per-tagged-thing copy in case the global thing goes away. We need that because the Taxonomies+Tags ownership and lifecycle is different from ObjectTags. The process of reconciliation is between per-object copies and the taxonomy's canonical Tag.

But do we really need that many copies? Can't we keep a more normalized representation as long as we create a copy that follows the same ownership as the ObjectTags in question?

What if we kept a copy of the Taxonomy at the layer of the importable/exportable thing (e.g. library, course, learning package)? Take a course for example. Let's say:

* ObjectTags don't have local copies of the name.
* All the ObjectTags for course content is against a course-level Taxonomy.
* Some of those are CourseTaxonomies that are entirely owned and internal to the course.
* Some of those are CourseTaxonomies that serve as links-with-copies to other Taxonomies (CourseTaxonomies controlled by other courses, OrgTaxonomies, SystemTaxonomies).
* Inconsistencies within the Course on import are errors (they can be ignored and the values dropped, or they can stop the process, but there's no reconciliation).
* Reconciliation for updated values happens _between Taxonomies_.

That would significantly cut down on the number of rows that would have to be iterated over during any sort of reconciliation process.

File: openedx_tagging/core/tagging/models.py
Comment (pomegranited) [reply openedx_tagging/core/tagging/models.py:7]: Hmm.. you're right.. maybe they'd be better in https://github.com/openedx/edx-django-utils/ ?

File: openedx_tagging/core/tagging/api.py
Comment (pomegranited) [reply openedx_tagging/core/tagging/api.py:68]: Hi @ormsbee , we haven't gone over an option like this, so thank you for raising it! And you're right, it is storage-heavy to store copies of the `name:value` on each ObjectTag, which will store the most rows of any model in this package.

But, this does seem more complicate, so do we care that much about storage space?

> Right now, it seems like we have the global representation of a Tag, and a per-tagged-thing copy in case the global thing goes away.

We have three uses for the `ObjectTag._value` field:

1. What you said ^
2. We can import tagged content without having to have a Taxonomy with these Tags in place before importing.
3. Free-text tags can be anything the content author wants, and are not hierarchical.
   So we don't store them as `Tag`s, we just create an `ObjectTag` with `_value="free text value"` and `tag=None`. This also avoids race conditions; we don't have to worry about creating both a `Tag` and an `ObjectTag` when free-text tagging an object.

(`ObjectTag._name` is also in the model for reasons 1 & 2.)

But there's a lot of dancing going on around this `_value` field, which likely means it's trying to do too much.

If I understand you correctly, you're suggesting something like:

* Taxonomy: canonical source of a taxonomy's tags and metadata. Taxonomy Admins create these, import/export tags.
* ContentTaxonomy: references a `source` Taxonomy`, and keeps a copy of its Tag tree, e.g. as ContentTaxonomyTag
* ContentTag: references a ContentTaxonomy and a ContentTaxonomyTag.
* ContentTaxonomy objects know how to reconcile changes made to its `source` Taxonomy or the source Taxonomy's Tag without losing `ContentTag` data.

So this deals with point #1 above, but how should we handle points #2 and #3?

File: openedx_tagging/core/tagging/models.py
Comment (bradenmacdonald) [reply openedx_tagging/core/tagging/models.py:50]: Other parts of the code that I saw recently definitely made that assumption.

File: openedx_tagging/core/tagging/models.py
Comment (bradenmacdonald) [inline openedx_tagging/core/tagging/models.py:406]: Should we have some kind of check here if this object is even allowed to be linked to this taxonomy? What if the taxonomy is for a completely different org than the course content that's tagged?

File: openedx_tagging/core/tagging/api.py
Comment (bradenmacdonald) [reply openedx_tagging/core/tagging/api.py:68]: My assumption was that reconciliation would be quite a rare process. If that's the case, I would strongly lean toward the simpler version, even if it uses a bit more space for storage. (Tags are usually quite small so I don't think the space is an issue.)

Even if it's not that rare, I think there's a simpler approach: If it's `ObjectTag.objects.all()` that's the problem, I suggest changing the reconciliation API to work on a single taxonomy or a single course at a time, rather than all tags of all objects. So for example it could take a course ID as its argument, and then do the reconciliation based on `ObjectTag.objects.filter(object_type=..., object_id__startswith="block-v1:the+course+ID@")`. Then you get much more efficient (and targeted) reconciliation at appropriate times (e.g. after a course import), and the data model is simple and easy to reason about.

What about taxonomy import? It seems like you might need to iterate over all the ObjectTags in that case, to see if any of them should be "upgraded" to new values in the taxonomy based on their `name` and `value`. But I think you can actually do a much more targeted iteration by only iterating the `ObjectTag` that have either (taxonomy_id is None) or (taxonomy_id==imported_taxonomy_id and tag_id is None). Which is again much much faster than iterating all the ObjectTags, because I expect that the vast majority of ObjectTags will be filtered out by that criteria (because they're linked to a taxonomy and/or a tag).

Comment (pomegranited) [reply openedx_tagging/core/tagging/api.py:68]: @bradenmacdonald 

> If that's the case, I would strongly lean toward the simpler version, even if it uses a bit more space for storage. 

That's my thinking too, and hence this PR.. but I am open to other options @ormsbee .

> Even if it's not that rare, I think there's a simpler approach: If it's ObjectTag.objects.all() that's the problem, I suggest changing the reconciliation API to work on a single taxonomy or a single course at a time, rather than all tags of all objects. 

👍 to adding filter options to the [resync_tags](https://github.com/open-craft/openedx-learning/blob/jill/taxonomy-api/openedx_tagging/core/tagging/api.py#L63-L73) method.

> What about taxonomy import? 

Perfect use of this, yes. I'll see if it'll fit into https://github.com/open-craft/openedx-learning/pull/1 , or create a follow-up task.

Thank you!

File: openedx_tagging/core/tagging/models.py
Comment (pomegranited) [reply openedx_tagging/core/tagging/models.py]: Oo good point.. enabled by [821ce62](https://github.com/openedx/openedx-learning/pull/57/commits/821ce62f8e6f127d97f9148b8bb88e0f29247ca1).

File: openedx_tagging/core/tagging/models.py
Comment (pomegranited) [reply openedx_tagging/core/tagging/models.py:50]: Fixed with [a11d1c9](https://github.com/openedx/openedx-learning/pull/57/commits/a11d1c9f6718cb5866b50ac7c9941823e49a5ecb).

**Comment by pomegranited** [comment]
Thank you for all your reviews and advice, @ChrisChV @ormsbee and @bradenmacdonald !

I think this is ready to merge now -- if you agree, could someone please squash+merge this? I haven't squashed and rebased the commits here to keep the conversations working, but I can do that if it's preferred.

File: openedx_tagging/core/tagging/api.py
Comment (ormsbee) [reply openedx_tagging/core/tagging/api.py]: My apologies: I thought I replied to this yesterday but I never submitted my comment.

@bradenmacdonald: Yes, my concern was much more about the number of records that need to be updated (and any potential locking), and not the space usage.

@pomegranited:

> 2. We can import tagged content without having to have a Taxonomy with these Tags in place before importing.

Yeah, what I was proposing assumes that we'd have local Taxonomies/Tags in place for import, because those taxonomies can be locally controlled at the scope the import is happening.

> 3. Free-text tags can be anything the content author wants, and are not hierarchical.
>     So we don't store them as `Tag`s, we just create an `ObjectTag` with `_value="free text value"` and `tag=None`. This also avoids race conditions; we don't have to worry about creating both a `Tag` and an `ObjectTag` when free-text tagging an object.

I'm fuzzy on this part. It doesn't have to block this PR, but don't we have to worry about race conditions where we're making the same free text value ObjectTag for the same object multiple times? It seems like it would benefit from being a separate model altogether, so we could put a uniqueness constraints around the freeform tag and the thing being tagged, independent of a taxonomy?

Comment (ormsbee) [reply openedx_tagging/core/tagging/api.py]: To be clear: I'm fine with the proposal to keep the general shape of the data structures now and do per-course filtering during reconciliation. We can always add in the local-taxonomy-copy thing later if it turns out we want it. The questions I'm asking now are just to improve my understanding.

**Comment by ormsbee** [comment]
@pomegranited: Could you please resolve merge conflicts? (I think it's fine for you to squash.)

File: openedx_tagging/core/tagging/models.py
Comment (pomegranited) [reply openedx_tagging/core/tagging/models.py:7]: Great plan, thank you @ormsbee . Should I codify this in an ADR in my next PR?

Comment (pomegranited) [reply openedx_tagging/core/tagging/models.py:7]: @ormsbee  These fields are working great on MySQL, but I can't use them in sqlite in the LMS tests :(

Running the tests with the branch from https://github.com/openedx/edx-platform/pull/32518 is throwing these errors when I run them on my [devstack](https://github.com/openedx/devstack)'s `lms-shell`:

```
root@lms:/edx/app/edxapp/edx-platform# pytest openedx/features/content_tagging/
<snip>
../venvs/edxapp/lib/python3.8/site-packages/pytest_django/plugin.py:490:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../venvs/edxapp/lib/python3.8/site-packages/pytest_django/fixtures.py:122: in django_db_setup
    db_cfg = setup_databases(
../venvs/edxapp/lib/python3.8/site-packages/django/test/utils.py:179: in setup_databases
    connection.creation.create_test_db(
../venvs/edxapp/lib/python3.8/site-packages/django/db/backends/base/creation.py:74: in create_test_db
    call_command(
../venvs/edxapp/lib/python3.8/site-packages/django/core/management/__init__.py:181: in call_command
    return command.execute(*args, **defaults)
../venvs/edxapp/lib/python3.8/site-packages/django/core/management/base.py:398: in execute
    output = self.handle(*args, **options)
../venvs/edxapp/lib/python3.8/site-packages/pytest_django/fixtures.py:276: in handle
    return super().handle(*args, **kwargs)
../venvs/edxapp/lib/python3.8/site-packages/django/core/management/base.py:89: in wrapped
    res = handle_func(*args, **kwargs)
../venvs/edxapp/lib/python3.8/site-packages/django/core/management/commands/migrate.py:262: in handle
    post_migrate_apps.render_multiple([
../venvs/edxapp/lib/python3.8/site-packages/django/core/management/commands/migrate.py:263: in <listcomp>
    ModelState.from_model(apps.get_model(*model)) for model in model_keys
../venvs/edxapp/lib/python3.8/site-packages/django/db/migrations/state.py:407: in from_model
    fields.append((name, field.clone()))
../venvs/edxapp/lib/python3.8/site-packages/django/db/models/fields/__init__.py:513: in clone
    return self.__class__(*args, **kwargs)
<snip>
ERROR openedx/features/content_tagging/tests/test_api.py::TestAPIContentTaxonomy::test_content_taxonomy_enabled_for_org1 - ValueError: Cannot use db_collation with MultiCollationCharField. Please use a db_collations dict instead.
```

And I found your post related to this on https://forum.djangoproject.com/t/supporting-collation-settings-on-multiple-database-backends/12561/, but I'm unclear how (or if) I can work around this.

Should I just roll these back to being normal, default-collation CharField/TextField until we work out a solution for both sqlite and MySQL?

File: openedx_tagging/core/tagging/api.py
Comment (pomegranited) [reply openedx_tagging/core/tagging/api.py]: > It seems like it would benefit from being a separate model altogether, so we could put a uniqueness constraints around the freeform tag and the thing being tagged, independent of a taxonomy?

I'll give it some thought.. originally when I was doing the discovery, I thought we'd end up having to subclass all of these base classes, which would lead to lots and lots of classes, e.g. SystemTaxonomy, SystemTaxonomyTag, SystemTaxonomyObjectTag, ...  And so I was trying to reduce the number of classes used at the base here. But now that I've gotten into the ContentTaxonomy implementation, I didn't even need to create a ContentTag class, because all of the ObjectTag validation is handled by the taxonomy. We'll still need a FreeFormTaxonomy of some sort, even if it has no tags, because that model is where the namespace for the object tags are stored.

Unfortunately we're running short on time for this phase 1, and so I'd rather not tweak it too much more until we actually have a frontend that's using it. But if I find that the frontend is having to do a lot of `if freeform then do this else do something else" before calling the backend, then separating them out will make much more sense.

File: openedx_tagging/core/tagging/models.py
Comment (pomegranited) [reply openedx_tagging/core/tagging/models.py:7]: If I remove the check ([a1a90d2](https://github.com/openedx/openedx-learning/pull/57/commits/a1a90d2689dd1537e2cc87544945d3abe0362e17)) then my tests on https://github.com/openedx/edx-platform/pull/32518 pass, cf https://github.com/openedx/edx-platform/actions/runs/5383371513/jobs/9770044365

Are you ok with merging this without that check?

Comment (ormsbee) [reply openedx_tagging/core/tagging/models.py:7]: Yes. I'm going to look through it later to try to figure out why it did this, but I'm fine with removing it for now. Thank you.

**Review by ormsbee — APPROVED** [review]

**Comment by ormsbee** [comment]
@pomegranited: is this okay for me to rebase+merge now?

**Comment by pomegranited** [comment]
@ormsbee Yes please! Sorry for the delay, I didn't think you were even up :)

**Comment by openedx-webhooks** [comment]
@pomegranited 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/openedx-learning&entry.867055334=https://github.com/openedx/openedx-learning/pull/57&entry.1484655318=https://github.com/pomegranited&entry.752974735=2023-06-13+08:10&entry.1917517419=2023-06-27+04:10&entry.2133058324=Yes) so we can improve your experience in the future.

**Comment by ormsbee** [comment]
I'm a parent. I keep weird hours sometimes. 😉

**Pull request merged**
